README.md
# YubiBridge

[Docs](https://defguard.gitbook.io/defguard/enterprise-features/yubikey-provisioning)

Yubi-Bridge is a Python module that creates [GPG](https://gnupg.org/) keys for
[YubiKey](https://www.yubico.com/products/) and transfers them automatically to YubiKey.
It can be run as a stand-alone application, or a client that takes jobs from DefGuard Core backend.
Refer to [documentation](https://defguard.gitbook.io/defguard/enterprise-features/yubikey-provisioning) for more info.

## How to run?

You can run this code:

* using poetry - recommended for development
* using docker-compose - recommended for regular use

### Using poetry:

> Make sure you have [Poetry](https://python-poetry.org/) installed.

```
# Clone the repository recursively (with protobuf files)
git clone --recursive git@github.com:DefGuard/yubi-bridge.git
cd yubi-bridge

# Install dependencies
poetry install

# Compile proto files
poetry run python3 -m grpc_tools.protoc -Iproto --python_out=yubi-bridge --grpc_python_out=yubi-bridge proto/worker/worker.proto

## Due to [issue](https://github.com/protocolbuffers/protobuf/issues/1491) with generated proto imports you need to replace import path manually

# Run YubiBridge
poetry run python3 -m yubi-bridge.main.py [options]
```

Example, running as a Defguard client:

```
poetry run python3 yubi-bridge/main.py --grpc defguard-grpc.mycompany.com --worker-token <YOUR_JWT_TOKEN> --id dev_worker
```

> You'll find the JWT token on "Provisioners" page of your Defguard instance.

Example, running as a stand-alone app:

```
poetry run python3 yubi-bridge/main.py --provision <first_name> <last_name> <email>
```

### Using docker-compose

Refer to [documentation](https://defguard.gitbook.io/defguard/enterprise-features/yubikey-provisioning#as-a-defguard-client)
for a guide on how to setup docker-compose.

## Environmental variables

You can set these variables to configure how the YubiBridge behaves.

| Environmental variable        | Default             | Description                                                                |
| ----------------------------- | ------------------- | -------------------------------------------------------------------------- |
| WORKER_ID                     | YubiBridge          | Worker id                                                                  |
| LOG_LEVEL                     | INFO                | Logging level, available levels are:    INFO,DEBUG,ERROR                   |
| URL                           | localhost:50055     | Url of your DefGuard instance                                              |
| JOB_INTERVAL                  | 2                   | Number of seconds in which worker ping DefGuard for data to provision      |
| SMARTCARD_RETRIES             | 1                   | Number of retries in case if there are no smartcards                       |
| SMARTCARD_RETRY_INTERVAL      | 15                  | Number of seconds before checking for smartcard again                      |


debian/changelog
yubi-bridge (0.1.0) stable; urgency=low

  * Initial release.

 -- Community <community@defguard.net>  Mon, 25 Aug 2023 12:00:00 +0000


yubi-bridge/tests/fixtures.py
correct_json = '{"first_name":"Jan","last_name":"Kowalski","email":"test@test.pl"}'
wrong_key_json = '{"frst_name":"Jan","ast_name":"Kowalski","mail":"test@test.pl"}'
wrong_format_json = '{first_name":"Jan","Last_name":Kowalski","email":"test@test.pl"}'

s_key = {
    "key_id": "BC256C4D535519B8",
    "key_type": "s",
    "key": "F1617E08EC3BE9502431AE66BC256C4D535519B8",
}
a_key = {
    "key_id": "968BA0FEBFD7ADA1",
    "key_type": "a",
    "key": "E2FA8E6611FCBCE113344CED968BA0FEBFD7ADA1",
}
e_key = {
    "key_id": "7BE87A524CAF0A1C",
    "key_type": "e",
    "key": "7654DB696EE79A17AB8395F47BE87A524CAF0A1C",
}

list_keys = [
    {
        "type": "pub",
        "trust": "u",
        "length": "2048",
        "algo": "1",
        "keyid": "795FA0E3332FCEFC",
        "date": "1637936835",
        "expires": "",
        "dummy": "",
        "ownertrust": "u",
        "sig": "",
        "cap": "escaESCA",
        "issuer": "",
        "flag": "",
        "token": "",
        "hash": "",
        "curve": "",
        "compliance": "23",
        "updated": "",
        "origin": "0",
        "uids": ["marek test <test@o2.pl>"],
        "sigs": [],
        "subkeys": [
            ["BC256C4D535519B8", "s", "F1617E08EC3BE9502431AE66BC256C4D535519B8"],
            ["968BA0FEBFD7ADA1", "a", "E2FA8E6611FCBCE113344CED968BA0FEBFD7ADA1"],
            ["7BE87A524CAF0A1C", "e", "7654DB696EE79A17AB8395F47BE87A524CAF0A1C"],
        ],
        "fingerprint": "93F8CF48897937A0FB2A0856795FA0E3332FCEFC",
        "subkey_info": {
            "BC256C4D535519B8": {
                "type": "sub",
                "trust": "u",
                "length": "2048",
                "algo": "1",
                "keyid": "BC256C4D535519B8",
                "date": "1637936836",
                "expires": "",
                "dummy": "",
                "ownertrust": "",
                "uid": "",
                "sig": "",
                "cap": "s",
                "issuer": "",
                "flag": "",
                "token": "",
                "hash": "",
                "curve": "",
                "compliance": "23",
                "updated": "",
                "origin": "unavailable",
            },
            "968BA0FEBFD7ADA1": {
                "type": "sub",
                "trust": "u",
                "length": "2048",
                "algo": "1",
                "keyid": "968BA0FEBFD7ADA1",
                "date": "1637936838",
                "expires": "",
                "dummy": "",
                "ownertrust": "",
                "uid": "",
                "sig": "",
                "cap": "a",
                "issuer": "",
                "flag": "",
                "token": "",
                "hash": "",
                "curve": "",
                "compliance": "23",
                "updated": "",
                "origin": "unavailable",
            },
            "7BE87A524CAF0A1C": {
                "type": "sub",
                "trust": "u",
                "length": "2048",
                "algo": "1",
                "keyid": "7BE87A524CAF0A1C",
                "date": "1637936839",
                "expires": "",
                "dummy": "",
                "ownertrust": "",
                "uid": "",
                "sig": "",
                "cap": "e",
                "issuer": "",
                "flag": "",
                "token": "",
                "hash": "",
                "curve": "",
                "compliance": "23",
                "updated": "",
                "origin": "unavailable",
            },
        },
    }
]

public_key_asc = """-----BEGIN PGP PUBLIC KEY BLOCK-----

mQENBGGg7sMBCADpbDfLjk1izSZSxBgQm6ZFBg5DgHy8CVsRaUiQlC4RvfjtPDSZ
9lihyACDS7znROY4m/F0Pm6fI4d71OohAEHOYcYRoGLdvDuUB6SQMM+8jwY5jIKE
q5fS207WUVg1L6sO0cb6mP/magDfsvJ/cPSRFins7Dk1E55jQlS0WzCyBrBWVeAf
6GAlbVHOMpXiSO/CBMGA9IVtGlM6WXwIH9WDjCQ12WPVZ0gzb8P7sc5dIfFYAZ1G
abGXUfbG5PoWPzHNLwxHLkXEFLpuyljDYrnw23AgVZGxqzQVj7aMahLQrD57Xckt
Ap6a1K1dHqjAbhjKrvWf/+8SYMOrS1PtwiQTABEBAAG0F21hcmVrIHRlc3QgPHRl
c3RAbzIucGw+iQFOBBMBCgA4FiEEk/jPSIl5N6D7KghWeV+g4zMvzvwFAmGg7sMC
Gy8FCwkIBwIGFQoJCAsCBBYCAwECHgECF4AACgkQeV+g4zMvzvx9AQf/eVWJNFkA
i1s14dBvHOB+30Lg07T3w8aUT9cimxeLvaZc4W5fcqoF1ZVaQnalNNScCTSznNPr
FA4vTrCPs9zrFUO5/o+EOsvb+7V9SqnfbyQqKmRX5RYOZxORf8v1jYiF07mg9ggH
yxP8cWgF4m0dJyNhcBbZlE6fHEO9HCoui743BpqQbD7jmIfdGZGNi8vxTRhByOU/
vDFPey569/urmE+hyAAgxmv0/129C5QkFsbOSth3e6ueS7jfViZU9sWtTgNYycik
jPc2pdzoFr1miCoSolbozU23R19lhdsfNKZ0JCDzFiYL8ZAssTqbtMPGjKJjrJRj
VT2Rs4mFjQFMtLkBDQRhoO7EAQgA1q8Ft/dauD6VtCt1ue5rPjZiBjUK2CGzS0/N
XDj5Cxph5UKG/dQNXWOoXoFDfv3M2vk1rszRbj0Fo8qcoeS8slMUogKSLiGD+n4E
CIAlTdivO1K5qG4RaFhtUOL6l1OT/Z/cMVrUGylfMbS7rVctUR3e/zHSWp4GHxO9
lXjxdRyI6J+hFbY7b/3ppcoGT4HPL6n0Xhblxt8mXyUvwanXtTNGhHS8rcvwJx/O
/EILLFzD+IT/qdS/AaFfDedMG6bb3TPcSBdqnuaZmf9ZJBR/gTnt/Et1SFX9eN7o
WAd0Zg45vieRLFgtHGplxZbcXZnpU05g5gd7BkHyhLY0IBCFQwARAQABiQJsBBgB
CgAgFiEEk/jPSIl5N6D7KghWeV+g4zMvzvwFAmGg7sQCGwIBQAkQeV+g4zMvzvzA
dCAEGQEKAB0WIQTxYX4I7DvpUCQxrma8JWxNU1UZuAUCYaDuxAAKCRC8JWxNU1UZ
uBFIB/9Qcv8txZwUByxkoPHaFhmPTrHL0WVBNDnnQRcQmZ11uuGvvvpMZfk2HCy3
QBhFI7t7S0lgf9VlhypEb3rILuiHxF7V4H0+/cwHTcbN9RdsfAsDIR1guItB7tW2
9zRMkKVo6wQaEXTO7nQf5JBzshecl2PJ8C0aCojM0DK5JGq6OBZ/H5ybeCf9SCxy
PxEk4e66BRnbmC3zpyIRQ7+pBnd2B+84FtLpkB0Ns0B/raPA8zLMnfPa63r29cDD
ok3gAzVvlEdCvCFNzNTRUSHsMLT1hG43RW9wsISshn7EdcDnvJybNqp9NfXYMKl2
ICEkmtqeSJTxtyaB+bE8joRvNXhBepQH/0g0bQyiGXxEPbsZHUkdQZEnOL3FXoUR
NX7mbqQDKVh73VuLlu/bJpk3zn0u5QQLCCCd3Bcm0WrgkRcofN29L4s8jlb7m2OP
xAHcrrJ3+7O46ksTQxWCDDqmup4eGjZaVXTC87LMzTfO1stru1nL+z2fUY8A5aDS
cdeHXPGSkEDiZkUWFd/V9AgzlUcS75K8cnWQTVJpZuK5+jS7ayettixr8myUzFZn
++4EpA+qJNV1A/QBZvpegsqyAT2n587aX3+Oy/xyzsXtE4bcwVcpPoGHy7tYIe4L
ijZO65VsD2fOtPm0Dbi5/A/amSr7TkFJg7k1vMBaoTz6FGTqxBNiHKC5AQ0EYaDu
xgEIALr44MYjFahbQ5szRX/lhjo638P3kj2HZQ+R67vyUR18G0tjb7blrh1OV7eE
MHo0X+sP0zB2OTRC2YiWxUia3beS7jXgIxYVg32UJSP2LWxfv5fN7vaOZ1VEv6Xc
/079XGdhatgpZLuRRD7AJE22hbHsl0zf2i/jzL1/zUUGNWwdfDmqcSLfx8JL4M8S
dNb/QhzWGFjXV51uHHygRVLzWJI8/cLciBFvEXeu+Lpxcg7DlgyylU0DSWy6Y1gc
FmRJlvYjKwDUjF0+rDBKaVufuEWU9CKfz0Q0a1SC2I40PtPz83fbFCvihTAgTF8j
jJ+eyzRLdFufvZ3Zaw2OVfdedf8AEQEAAYkBNgQYAQoAIBYhBJP4z0iJeTeg+yoI
VnlfoOMzL878BQJhoO7GAhsgAAoJEHlfoOMzL878kNoIAJcW4UBzJ/l2xa/nbtK9
QeNTi5/qLaG0dx6k5gHHgJjOdi0+j7bEWqhqsvvKrkzteRGwYol3dx6dnpxpoo7D
MNzpDLuUSjnwpldnP9HgVkHT7Iw4HT0OMQjV19axp9/woStxpgxr5YirWU/a4OU1
z4aAHgnAHNFc1pAXGoaZnc1TQi/YtQmmDkDpJQnAGbh4r520ZWxdei12soTOQfnS
OVSludy88e8RZwamCgG0uubIjk6rZbNkea+r0v+4+ZhxoxZUtSPWM9xiKRmEC3YY
4um6h7pcN2bcyJBHTmK6EwRP+vaHXnZqa2eWUzcAdQ6onQzOUnNGJejBLxBjHkkw
Xuy5AQ0EYaDuxwEIAMU00CxDawhc2mf4To5pmlLti++bPBSr+lJSQNwUbyU51vlS
TUf+qEAoFdsdzYvbXnngyxz3ibSQEHd0mdMmjA7mWFviLb2kVzFB5np/m+SO/70M
LJr7MYcgBltNnUsm06fumw2P4NGHH5hS1x5OqIZFDnr0x8jDdNkTRvzxZVf3rlDZ
KIb9YaMQHaXAecoK715AksouR80ehHcE6DvWnfleJxytv65eNDVTTQi4+SzzVday
AJxfvgf4P9HOCFwjnd6pl0rJWP1dm8liJm1I/vQchuG9oO5YMio0HCFyV8CldBUr
VV8SqNH1ImjjTO2VKnAWxvRN8D9wy0sVBm/gYEEAEQEAAYkBNgQYAQoAIBYhBJP4
z0iJeTeg+yoIVnlfoOMzL878BQJhoO7HAhsMAAoJEHlfoOMzL878IRwIAL8quK8D
GLQD0efi+TBcdP2liSr1AThkXFOEJr6hV6g1h1Ih3Cdzqz23NoGl20IFStGlW3qZ
8PE9GLdik2YP8uzT8UCkEnzDHmLdmohXOAnfUt7BknpPt1wQYPw9cod53iU4cSMO
uLXsxNpQg0oFVaBHS3T8xpeXnjOKSP/gS6zBHY2iyssyLtizYaGGIobyDUtrDY41
kEMNHyKfqw3Xxw+ezx8b79Sa6W2YyY2raSAhP2UJFI/S7A85P9rC6tABTBswilq9
yFexCwnQoolzUrBZbj7yeK/vuFZaxfMXf2/ag66f4ny4z4TiPg+Xpqch7EtG92LC
i7u4BVRJEgtqV04=
=3+aj
-----END PGP PUBLIC KEY BLOCK-----
"""


yubi-bridge/tests/test_main.py
from tempfile import TemporaryDirectory

import pytest

from main import YubiBridge
from tests import fixtures


# Succes write to yubikey
@pytest.fixture
def mock_write_to_yubikey(mocker):
    mocker.patch("YubiBridge.write_to_yubikey", return_value=None)


@pytest.fixture
def mock_write_to_yubikey_error(mocker):
    mocker.patch("YubiBridge.write_to_yubikey", return_value=SystemExit)


@pytest.fixture
def mock_create_keys(mocker):
    mocker.patch("YubiBridge.create_keypair", return_value="FINGERPRINT")


@pytest.fixture
def mock_yubikey_success(mocker):
    mocker.patch("YubiBridge.check_yk", return_value=True)


@pytest.fixture
def mock_nodevices(mocker):
    mocker.patch("ykman.device.list_all_devices", return_value=[])


@pytest.fixture
def mock_pubkeys(mocker):
    mocker.patch("gnupg.GPG.list_keys", return_value=fixtures.list_keys)
    mocker.patch("gnupg.GPG.export_keys", return_value=fixtures.public_key_asc)


def test_create_keys():
    with TemporaryDirectory() as tempdir:
        yb = YubiBridge(tempdir)
        assert yb.create_keypair("marek test", "test@o2.pl")


yubi_bridge/tests/test_main.py
from tempfile import TemporaryDirectory

import pytest

from yubi_bridge.main import YubiBridge
from . import fixtures


# Succes write to yubikey
@pytest.fixture
def mock_write_to_yubikey(mocker):
    mocker.patch("YubiBridge.write_to_yubikey", return_value=None)


@pytest.fixture
def mock_write_to_yubikey_error(mocker):
    mocker.patch("YubiBridge.write_to_yubikey", return_value=SystemExit)


@pytest.fixture
def mock_create_keys(mocker):
    mocker.patch("YubiBridge.create_keypair", return_value="FINGERPRINT")


@pytest.fixture
def mock_yubikey_success(mocker):
    mocker.patch("YubiBridge.check_yk", return_value=True)


@pytest.fixture
def mock_nodevices(mocker):
    mocker.patch("ykman.device.list_all_devices", return_value=[])


@pytest.fixture
def mock_pubkeys(mocker):
    mocker.patch("gnupg.GPG.list_keys", return_value=fixtures.list_keys)
    mocker.patch("gnupg.GPG.export_keys", return_value=fixtures.public_key_asc)


def test_create_keys():
    with TemporaryDirectory() as tempdir:
        yb = YubiBridge(tempdir)
        assert yb.create_keypair("marek test", "test@o2.pl")


yubi_bridge/client.py
import collections
import logging

import grpc

from .worker.worker_pb2 import JobStatus, Worker
from .worker.worker_pb2_grpc import WorkerServiceStub


class _ClientCallDetails(
    collections.namedtuple(
        "_ClientCallDetails", ("method", "timeout", "metadata", "credentials")
    ),
    grpc.ClientCallDetails,
):
    pass


class WorkerTokenInterceptor(grpc.UnaryUnaryClientInterceptor):
    """Inject secret token to authenticate the worker."""

    def __init__(self, secret: str):
        self.secret = secret

    def intercept_unary_unary(self, continuation, client_call_details, request):
        new_details = self.inject_token(client_call_details)
        response = continuation(new_details, request)
        return response

    def inject_token(self, client_call_details):
        if client_call_details.metadata is None:
            metadata = []
        else:
            metadata = list(client_call_details.metadata)
        metadata.append(
            (
                "authorization",
                self.secret,
            )
        )
        client_call_details = _ClientCallDetails(
            client_call_details.method,
            client_call_details.timeout,
            metadata,
            client_call_details.credentials,
        )
        return client_call_details


class WorkerHandler:
    """Handle communication with gRPC server."""

    def __init__(self, worker_id: str, address: str, secret: str, secure=False):
        if secure:
            channel_credential = grpc.ssl_channel_credentials()
            channel = grpc.secure_channel(address, channel_credential)
        else:
            channel = grpc.insecure_channel(address)
        self.channel = grpc.intercept_channel(channel, WorkerTokenInterceptor(secret))
        self.worker_id = worker_id

    def register(self):
        stub = WorkerServiceStub(self.channel)
        request = Worker(id=self.worker_id)
        logging.debug(request)
        try:
            response = stub.RegisterWorker(request)
            logging.debug(response)
        except grpc.RpcError as err:
            if err.code() != grpc.StatusCode.ALREADY_EXISTS:
                logging.error("Server error: %s %s", err.code(), err.details())

    def get_job(self) -> tuple:
        stub = WorkerServiceStub(self.channel)
        request = Worker(id=self.worker_id)
        try:
            response = stub.GetJob(request)
            return (
                response.first_name,
                response.last_name,
                response.email,
                response.job_id,
            )

        except grpc.RpcError as err:
            if err.code() != grpc.StatusCode.NOT_FOUND:
                logging.error("Server error: %s %s", err.code(), err.details())

        return ()

    def send_job_status(
        self,
        job_id: int,
        success: bool,
        public_key: str,
        ssh_key: str,
        fingerprint: str,
        error: str,
    ):
        stub = WorkerServiceStub(self.channel)
        request = JobStatus(
            id=self.worker_id,
            job_id=job_id,
            success=success,
            public_key=public_key,
            ssh_key=ssh_key,
            fingerprint=fingerprint,
            error=error,
        )
        try:
            stub.SetJobDone(request)
            if success:
                logging.info("Keys sent to DG")
        except grpc.RpcError as err:
            logging.error("Server error: %s %s", err.code(), err.details())


yubi_bridge/main.py
import argparse
import logging
import re
from os import getenv
from pathlib import Path
from subprocess import Popen, PIPE, run, CalledProcessError
from tempfile import TemporaryDirectory
from time import sleep
from typing import Tuple

import gnupg
import pexpect
from ykman.device import list_all_devices

from .client import WorkerHandler

URL = getenv("URL", "localhost:50055")
WORKER_ID = getenv("WORKER_ID", "YubiBridge")
DEFGUARD_TOKEN = getenv("DEFGUARD_TOKEN", "DEFGUARD_TOKEN")
LOG_LEVEL = getenv("LOG_LEVEL", logging.INFO)
SMARTCARD_RETRIES = getenv("RETRIES", 1)
JOB_INTERVAL = getenv("JOB_INTERVAL", 2)
SMARTCARD_RETRY_INTERVAL = getenv("SMARTCARD_INTERVAL", 15)

LOG = logging.getLogger(__name__)


class YubiBridge:
    def __init__(self, command: str = None):
        """Create homedir and initialise GPG with it."""
        # Try to kill default gpg-agent, which might interfere with our homedir.
        comp_proc = run(
            ["gpgconf", "--kill", "gpg-agent"],
            capture_output=True,
            check=True,
            text=True,
        )
        if comp_proc.stderr:
            LOG.error(comp_proc.stderr)
        if comp_proc.stdout:
            LOG.debug(comp_proc.stdout)

        self.tempdir = TemporaryDirectory()
        self.homedir = self.tempdir.name
        LOG.debug("Temporary directory %s", self.homedir)
        self.gpg = gnupg.GPG(gnupghome=self.homedir)
        self.gpg.encoding = "utf-8"
        # TODO: probably not needed when master key is going to be discarded
        self.passphrase = "password"
        self.pin = "123456"  # default PIN
        self.admin_pin = "12345678"  # default admin PIN
        self.command = command

    def keys_exist(self) -> bool:
        """Checks if GnuPG keychain contains any keys."""
        return bool(self.gpg.list_keys())

    def delete_keypair(self, fingerprint):
        """Delete private and public key for a given fingerprint."""
        LOG.debug("Deleting private key %s", fingerprint)
        self.gpg.delete_keys(fingerprint, True, passphrase=self.passphrase)
        LOG.debug("Deleting public key %s", fingerprint)
        self.gpg.delete_keys(fingerprint)

    def delete_all_keys(self):
        """Delete all keys from GnuPG keychain."""
        for key in self.gpg.list_keys():
            self.delete_keypair(key["fingerprint"])

    def create_keypair(self, name: str, email: str) -> str:
        """Create private/public key pair. Return key fingerprint."""
        # Generate master key for signing and a sub-key for encryption.
        LOG.info("Creating keys for %s %s", name, email)
        input_data = self.gpg.gen_key_input(
            key_type="RSA",
            key_length=4096,
            key_usage="sign,cert",  # encr, sign, auth, cert
            name_real=name,
            name_email=email,
            passphrase=self.passphrase,
            expire_date=0,  # key does not expire
            subkey_type="RSA",
            subkey_length=4096,
            subkey_usage="encr",
        )

        LOG.debug("Creating master key")
        master_key = self.gpg.gen_key(input_data)

        # Generate sub-key for authentication
        # TODO: this fails with 'TypeError: expected str, bytes or
        # os.PathLike object, not GenKey'
        # subkey = self.gpg.add_subkey(master_key,
        #  master_passphrase=self.passphrase, usage="auth")
        args = [
            self.gpg.gpgbinary,
            "--homedir",
            self.homedir,
            "--pinentry-mode=loopback",
            "--no-tty",
            "--passphrase",
            self.passphrase,
            "--quick-add-key",
            master_key.fingerprint,
            "rsa4096",
            "auth",
        ]
        LOG.debug("Running %s", " ".join(args))
        comp_proc = run(args, capture_output=True, check=True, text=True)
        if comp_proc.stderr:
            LOG.error(comp_proc.stderr)
        if comp_proc.stdout:
            LOG.debug(comp_proc.stdout)

        # sanity check
        expected_subs = [
            key
            for key in self.gpg.list_keys()
            if key["fingerprint"] == master_key.fingerprint
        ][0]["subkeys"]
        assert len(expected_subs) == 2

        LOG.debug("Created key %s", master_key.fingerprint)
        return master_key.fingerprint

    def public_key(self, fingerprint: str) -> str:
        """Return public key for a given fingerprint."""
        # KeyID is last 16 hex digits of fingerprint.
        return self.gpg.export_keys(fingerprint[-16:])

    def private_key(self, fingerprint: str) -> str:
        """Return private key for a given fingerprint."""
        # KeyID is last 16 hex digits of fingerprint.
        return self.gpg.export_keys(fingerprint[-16:], True, passphrase=self.passphrase)

    def ssh_key(self, fingerprint: str) -> str:
        """Return public key for in SSH format."""
        # FIXME: not implemented in python-gnupg
        args = [
            self.gpg.gpgbinary,
            "--homedir",
            self.homedir,
            "--export-ssh-key",
            fingerprint[-16:],
        ]
        LOG.debug("Running %s", " ".join(args))
        comp_proc = run(args, capture_output=True, check=True, text=True)
        if comp_proc.stderr:
            LOG.error(comp_proc.stderr)
        return comp_proc.stdout

    def check_connection(self):
        """Check if any smartcard is connected"""
        self.restart_pcscd()
        LOG.debug("Checking for smartcard...")
        try:
            for _, info in list_all_devices():
                LOG.debug("Found device with serial: %s", info.serial)
            return len(list_all_devices()) != 0
        except Exception:
            LOG.error("Device error")

    def keys_to_card(self, email: str) -> Tuple[bool, str]:
        """Transfer keys to card. Return True on success."""
        # Note: there is no way to point GnuPG to a desired card.
        # expected prompts
        prompt_bool = "\r\n[GNUPG:] GET_BOOL keyedit.keytocard.use_primary\r\n"
        prompt_line = "\r\n[GNUPG:] GET_LINE keyedit.prompt\r\n"
        prompt_pass = "\r\n[GNUPG:] GET_HIDDEN passphrase.enter\r\n"
        prompt_repl = "\r\n[GNUPG:] GET_BOOL cardedit.genkeys.replace_key\r\n"
        prompt_okay = "\r\n[GNUPG:] GET_BOOL keyedit.save.okay\r\n"
        prompt_type = "\r\n[GNUPG:] GET_LINE cardedit.genkeys.storekeytype\r\n"
        LOG.info("Transfering keys to card")
        cmd = pexpect.spawn(
            self.gpg.gpgbinary,
            [
                "--homedir",
                self.homedir,
                "--command-fd=0",
                "--status-fd=2",
                "--batch",
                "--pinentry-mode=loopback",
                "--edit-key",
                email,
            ],
            encoding="utf-8",
            logfile=open("/tmp/yb-log.txt", "w", encoding="utf-8"),
        )
        for prompt, line in (
            # store primary key
            (prompt_line, "keytocard"),
            (prompt_bool, "y"),
            (prompt_type, "1"),
            (prompt_pass, self.passphrase),
            (prompt_pass, self.admin_pin),
            (prompt_pass, self.admin_pin),
            # store encryption key
            (prompt_line, "key 1"),
            (prompt_line, "keytocard"),
            (prompt_type, "2"),
            (prompt_pass, self.passphrase),
            (prompt_pass, self.admin_pin),
            (prompt_line, "key 1"),
            # store authentication key
            (prompt_line, "key 2"),
            (prompt_line, "keytocard"),
            (prompt_type, "3"),
            (prompt_pass, self.passphrase),
            (prompt_pass, self.admin_pin),
            # good-bye
            (prompt_line, "q"),
            (prompt_okay, "y"),
        ):
            index = cmd.expect_exact(
                [prompt, prompt_repl, pexpect.EOF, pexpect.TIMEOUT]
            )
            if index == 1:
                # WARNING: such a key has already been stored on the card!
                err = "Card already has the key"
                LOG.error(err)
                LOG.error(
                    "Reset card to factory settings first with command gpg --card-edit"
                )
                return False, err
            if index == 2:
                err = "GnuPG process disconnected"
                LOG.error(err)
                return False, err
            if index == 3:
                err = "Can't connect to smartcard"
                LOG.error(err)
                return False, err
            cmd.sendline(line)
        # wait for graceful exit, or card info won't be saved
        cmd.expect_exact([pexpect.EOF, pexpect.TIMEOUT])
        LOG.info("Keys have been transferred to card")

        return True, "None"

    def check_yubikey(self) -> bool:
        """Check if keys were added to YubiKey."""
        with Popen(
            [self.gpg.gpgbinary, "--homedir", self.homedir, "--card-status"],
            stdout=PIPE,
        ) as p:
            sout = p.communicate()
            sout = sout[0].decode()
            try:
                result = re.search("(?<=General key info..: )(.*)", sout).group(0)
                if result != "[none]":
                    return True
                print(
                    "Something went wrong.."
                    "Reset yubikey to factory settings and try again"
                )
            except AttributeError:
                print("No gpg keys")

        return False

    def restart_pcscd(self) -> None:
        """Restart smartcard driver if exists"""
        path = Path("/etc/init.d/pcscd")
        if path.exists():
            args = [path, "restart"]
            cmd = run(args, capture_output=True, check=True, text=True)
            LOG.debug(cmd.stdout)

    def run_command(
        self, command: str, firstname: str, lastname: str, ssh_key: str, public_key: str
    ) -> None:
        args = [*command, firstname, lastname, ssh_key, public_key]
        LOG.debug("Running %s", " ".join(args))
        try:
            comp_proc = run(
                args,
                capture_output=True,
                check=True,
                text=True,
            )
            if comp_proc.stderr:
                LOG.error(comp_proc.stderr)
            if comp_proc.stdout:
                LOG.debug(comp_proc.stdout)
        except CalledProcessError as e:
            LOG.error(e.stdout)
            LOG.error(e.stderr)
        LOG.info("Command: %s have finished", " ".join(command))

    def provision(
        self, firstname: str, lastname: str, email: str
    ) -> Tuple[bool, str, str, str, str]:
        """
        Do entire provisioning.
        Return public key in PGP and SSH formats, and key fingerprint.
        """
        self.restart_pcscd()
        fullname = f"{firstname} {lastname}"
        fingerprint = self.create_keypair(fullname, email)
        public_key = self.public_key(fingerprint)
        ssh_key = self.ssh_key(fingerprint)
        # TODO: --card-edit
        success, err = self.keys_to_card(email)
        if success and self.command:
            self.run_command(self.command, firstname, lastname, ssh_key, public_key)
        self.delete_keypair(fingerprint)
        return (success, public_key, ssh_key, fingerprint, err)


def main():
    """Main program entry."""
    parser = argparse.ArgumentParser(
        prog="yubikey-bridge",
        description="DefGuard YubiKey provisioning tool",
    )
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument(
        "-p",
        "--provision",
        nargs=3,
        metavar=("<firstname>", "<lastname>", "<email>"),
        help="Provision with the following data",
    )
    group.add_argument(
        "-g",
        "--grpc",
        nargs="?",
        const=f"{URL}",
        help="Connect to gRPC server",
    )
    parser.add_argument(
        "-i",
        "--id",
        default=WORKER_ID,
        help="Worker ID (default: YubiBridge)",
    )
    parser.add_argument("-d", "--debug", action="store_true", help="Enable debug mode")
    parser.add_argument(
        "-s",
        "--secure",
        action="store_true",
        help="Enable secure (TLS) connection to gRPC server",
    )
    parser.add_argument(
        "-w",
        "--worker-token",
        default=DEFGUARD_TOKEN,
        help="Secret worker token to secure gRPC communication",
    )
    parser.add_argument(
        "-c",
        "--command",
        nargs="+",
        help="Run command using keys as arguments",
    )
    args = parser.parse_args()

    logging.basicConfig(
        level=logging.DEBUG if args.debug else LOG_LEVEL,
        format="%(asctime)s - %(levelname)s: %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )

    yb = YubiBridge(args.command)
    if yb.keys_exist():
        LOG.error("Some keys exist in GnuPG tempdir.")
        return

    if args.provision is not None:
        yb.provision(args.provision[0], args.provision[1], args.provision[2])

    if args.grpc is not None:
        LOG.info("Connecting to %s as %s", args.grpc, args.id)
        worker = WorkerHandler(args.id, args.grpc, args.worker_token, args.secure)
        worker.register()
        LOG.info("Worker connected")
        while True:
            job = worker.get_job()
            if job:
                LOG.info(
                    "Received job with id %s to provision yubikey for: %s %s %s",
                    job[3],
                    job[0],
                    job[1],
                    job[2],
                )
                for i in range(int(SMARTCARD_RETRIES) + 1):
                    if yb.check_connection():
                        success, public_key, ssh_key, fingerprint, err = yb.provision(
                            job[0], job[1], job[2]
                        )
                        LOG.info("Sending job status to DefGuard")
                        worker.send_job_status(
                            job[3], success, public_key, ssh_key, fingerprint, err
                        )
                        break

                    LOG.error(
                        "No smartcard found, please insert YubiKey "
                        "or if YubiKey is inserted, re-insert it."
                    )

                    if i == SMARTCARD_RETRIES:
                        LOG.info("Sending job status to DefGuard")
                        worker.send_job_status(
                            job[3],
                            False,
                            "-",
                            "-",
                            "-",
                            "Can't connect to smartcard",
                        )
                        break

                    LOG.info(
                        "Waiting %s seconds before looking for smartcard again",
                        SMARTCARD_RETRY_INTERVAL,
                    )
                    sleep(int(SMARTCARD_RETRY_INTERVAL))
            sleep(int(JOB_INTERVAL))


if __name__ == "__main__":
    main()
